pyphare/pyphare_tests/test_pharesee/test_geometry.py

```diff
@@ -0,0 +1,528 @@
+        self.assertTrue(touch_domain_border(Box(40, 64), self.hierarchy.domain_box, "upper"))
+
+
+
+
+    def test_particle_ghost_area_boxes(self):
+
+        expected = {
+
+            0: [ {"boxes":[Box(33, 33), Box(-1,-1)]},
+                 {"boxes": [Box(32, 32), Box(65, 65)]} ],
+
+
+            1: [ {"boxes": [Box(9, 9), Box(60, 60)]},
+                 {"boxes": [Box(63, 63), Box(112, 112)]} ]
+        }
+
+        gaboxes = particle_ghost_area_boxes(self.hierarchy)
+
+        # same number of levels
+        self.assertEqual(len(expected), len(gaboxes))
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+
+            # same number of PatchDatas
+            self.assertEqual(len(gaboxes[ilvl]), len(expected[ilvl]))
+
+            for act_pdata, exp_pdata in zip(gaboxes[ilvl], expected[ilvl]):
+
+                    self.assertEqual(len(exp_pdata["boxes"]), len(act_pdata["boxes"]))
+
+                    for exp_box in exp_pdata["boxes"]:
+                        self.assertTrue(exp_box in act_pdata["boxes"])
+
+
+
+
+
+    def test_level_ghost_boxes(self):
+
+        expected = {
+
+            1: [
+                {"boxes": [Box(9, 9),]},
+                {"boxes": [Box(60, 60),] },
+                {"boxes": [Box(63, 63),]},
+                {"boxes": [Box(112, 112),]}
+            ]
+        }
+
+        lvl_gaboxes = level_ghost_boxes(self.hierarchy)
+        for ilvl  in range(1, len(self.hierarchy.patch_levels)):
+            for actual, exp in zip(lvl_gaboxes[ilvl], expected[ilvl]):
+
+                act_boxes = actual["boxes"]
+                exp_boxes = exp["boxes"]
+
+                for act_box, exp_box in zip(act_boxes, exp_boxes):
+                    self.assertEqual(act_box, exp_box)
+
+
+
+
+
+    def test_overlapped_fields_are_equal(self):
+
+        overlaps = hierarchy_overlaps(self.hierarchy)
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+
+            for overlap in overlaps[ilvl]:
+
+                pd1, pd2 = overlap["pdatas"]
+                box      = overlap["box"]
+                offsets  = overlap["offset"]
+
+                self.assertEqual(pd1.quantity, pd2.quantity)
+
+                if pd1.quantity == 'field':
+
+                    # we need to transform the AMR overlap box, which is thus
+                    # (because AMR) common to both pd1 and pd2 into local index
+                    # boxes that will allow to slice the data
+
+                    # the patchData ghost box that serves as a reference box
+                    # to transfrom AMR to local indexes first needs to be
+                    # shifted by the overlap offset associated to it
+                    # this is because the overlap box has been calculated from
+                    # the intersection of possibly shifted patch data ghost boxes
+
+                    loc_b1 = boxm.amr_to_local(box, boxm.shift(pd1.ghost_box, offsets[0]))
```

Comment:
ok, the 2 offsets are 1 for each patch data. ah... and offset depends on the patchdata because the patch is defined by the cell indexes, while the patch data depends on the quantity... and primal or dual fields will thus not necessarily have the same offset. is that right ?

Reply:
not really, we'll discuss it tomorrow over the phone, I already explained a bit in a previous answer if you want to re-read those.